### PR TITLE
New survey builder

### PR DESCRIPTION
**Pull request by rmjjjt**

### CODE REVIEW

File: module/Application/view/application/builder/index.phtml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+            <label for="survey-name"><h3>Survey Name</h3></label>
+            <input type="text" id="survey-name" name="survey-name">
+        </form>
+    </div>
+</div>
+<div class="survey-section">
+    <h3>Questions</h3>
+    <div id="question-container"></div>
+</div>
+<hr>
```
Comment (aimeeew) [inline module/Application/view/application/builder/index.phtml:16]: css instead?

File: module/Application/view/application/builder/index.phtml [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+            <option value="text-input" selected>Text</option>
+            <option value="radio-input">Radio</option>
+            <option value="checkbox-input">Checkbox</option>
+        </select>
+        <label for="required">Required?</label>
+        <input type="checkbox" id="required" name="required">
+    </div>
+    <input id="add-question" class="btn btn-sm btn-success" type="submit" value="Add">
+</div>
+<hr>
```
Comment (aimeeew) [inline module/Application/view/application/builder/index.phtml:34]: this one too?

File: public/js/surveyBuilderForm.js [hunk 3]
```diff
@@ -0,0 +1,190 @@
+$(function(){
+    var addQuestionButton = $('#add-question')
```
Comment (aimeeew) [inline public/js/surveyBuilderForm.js]: $variable for jQuery selector?

File: public/js/surveyBuilderForm.js [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+
+    addQuestionButton.prop("disabled",true)
+
+    $('#input-selector').change(function()
+    {
+        if ($('#input-selector').val() !== 'text-input' && $('#question-options').length < 1)
+        {
+            addOptionsCreator($('#input-container'))
+
+        } else if($('#input-selector').val() == 'text-input')
```
Comment (aimeeew) [inline public/js/surveyBuilderForm.js]: could this just be an else?

File: public/js/surveyBuilderForm.js [hunk 5]
```diff
@@ -0,0 +1,190 @@
+$(function(){
+    var addQuestionButton = $('#add-question')
+
+    addQuestionButton.prop("disabled",true)
+
+    $('#input-selector').change(function()
```
Comment (aimeeew) [inline public/js/surveyBuilderForm.js]: lots of repetition of this input-selector - could be a variable?

File: public/js/surveyBuilderForm.js [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+            addOptionsCreator($('#input-container'))
+
+        } else if($('#input-selector').val() == 'text-input')
+        {
+            $('#question-options').remove()
+        }
+        validateNewQuestion($('#question').val(), addQuestionButton, $('#input-selector').val())
+    })
+
+    $('#question').keyup(function(e)
```
Comment (aimeeew) [inline public/js/surveyBuilderForm.js]: do you need 'e' here if it's not used?

File: public/js/surveyBuilderForm.js [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+    {
+        validateNewQuestion($('#question').val(), addQuestionButton, $('#input-selector').val())
+    })
+
+    addQuestionButton.click(function(e)
```
Comment (aimeeew) [inline public/js/surveyBuilderForm.js]: e

File: public/js/surveyBuilderForm.js [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+            if($required.is(':checked')){
+                question += ' *'
+            }
+
+            if(type == 'text') {
+                response += '<input type="text" disabled >'
+            } else {
+                if(options.length)
+                {
+                    options.each(function(key,option)
```
Comment (aimeeew) [inline public/js/surveyBuilderForm.js]: space in parameters?

File: public/js/surveyBuilderForm.js [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+            var response = '<div class="options">'
+
+            if($required.is(':checked')){
+                question += ' *'
+            }
+
+            if(type == 'text') {
+                response += '<input type="text" disabled >'
+            } else {
+                if(options.length)
```
Comment (aimeeew) [inline public/js/surveyBuilderForm.js]: would this work as an else if?

File: public/js/surveyBuilderForm.js [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+                    options.each(function(key,option)
+                    {
+                        response += '<div><input type="' + type + '" disabled value="' + option.value + '"> ' + option.value + '</div>'
+                    })
+                }
+            }
+
+            response += '</div>'
+
+            var newQuestion =   '<div class="new-question ui-state-default">' +
```
Comment (aimee-wyld) [inline public/js/surveyBuilderForm.js]: spacing?

File: public/js/surveyBuilderForm.js [hunk 11]
```diff
@@ -0,0 +1,190 @@
+$(function(){
```
Comment (aimeeew) [inline public/js/surveyBuilderForm.js]: some comments throughout could be useful?

File: public/js/surveyBuilderForm.js [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+    })
+}
+
+
+/**
+ * Creates a disabled field containing the text passed in, with a remove (-) button .
+ *
+ * @param $optionContainer OBJECT the div containing the text input and add button for creating new options
+ */
+function  createOption($optionContainer)
```
Comment (aimeeew) [inline public/js/surveyBuilderForm.js]: extra space?

File: public/js/surveyBuilderForm.js [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+
+/**
+ * Removes the selected option from the list and then validates the form.
+ *
+ * @param currentOption OBJECT the remove button for the option to be removed
+ */
+function removeOption(currentOption)
+{
+    $(currentOption).parent('div').remove()
+    validateNewQuestion($('#question').val(), $('#add-question'), $('#input-selector').val())
```
Comment (aimeeew) [inline public/js/surveyBuilderForm.js]: these same selectors are used throughout?

**Review by aimeeew — CHANGES_REQUESTED** [review]

File: public/js/surveyBuilderForm.js (hunk 5 above)
Comment (ralph-rutter) [reply public/js/surveyBuilderForm.js]: Done.

File: public/js/surveyBuilderForm.js (hunk 12 above)
Comment (ralph-rutter) [reply public/js/surveyBuilderForm.js]: It looks weird when the function description is only on one line, but this style of having a blank line between the description and parameter list is consistent across my DocBlocks and seems (after a little googling) to be a style adopted by others.

File: public/js/surveyBuilderForm.js (hunk 4 above)
Comment (ralph-rutter) [reply public/js/surveyBuilderForm.js]: I thought the same initially, as did Mike, but there is a slight distinction: 

The "if" checks if the input selector is being changed from the text type to another type, in which  case the options controller pops up.
The "else if" checks if the new type is text, in which case it removes the options controls. The case where it is being changed from radio to checkbox or vice versa is false in both cases, and hence we neither create nor remove the options controls.

File: public/js/surveyBuilderForm.js (hunk 12 above)
Comment (aimeeew) [reply public/js/surveyBuilderForm.js]: Sorry my comment was really unclear but we meant that before the function name 'createOption' it looks like there's more than one space (super picky but we were trying to be really careful!)

File: public/js/surveyBuilderForm.js (hunk 4 above)
Comment (aimeeew) [reply public/js/surveyBuilderForm.js]: That makes sense - thanks for explaining :)

File: public/js/surveyBuilderForm.js (hunk 13 above)
Comment (ralph-rutter) [reply public/js/surveyBuilderForm.js]: I had a look and could see no reason why they couldn't be variables, so started making the changes. I then remembered a chat I had with Mike yesterday about this and he said this is one of those development compromises: on one hand it's inefficient to keep making jQuery selections, but on the other validateNewQuestion() is called by other, unrelated, functions, so in order to enforce dependency injection the variables would have to be passed into places where it doesn't really make sense to. He concluded that it's better to leave as is. I'll go and undo the changes I started making then! lol

File: public/js/surveyBuilderForm.js (hunk 3 above)
Comment (ArcticFoxJ) [reply public/js/surveyBuilderForm.js]: fixed

File: public/js/surveyBuilderForm.js (hunk 7 above)
Comment (ArcticFoxJ) [reply public/js/surveyBuilderForm.js]: removed

File: public/js/surveyBuilderForm.js (hunk 9 above)
Comment (ArcticFoxJ) [reply public/js/surveyBuilderForm.js]: actually the if is no longer necessary so changed to just an else

File: public/js/surveyBuilderForm.js (hunk 8 above)
Comment (ArcticFoxJ) [reply public/js/surveyBuilderForm.js]: done

File: public/js/surveyBuilderForm.js [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+    $type.change(function () { // attaching a listener for the question input type being changed
+
+        //checking that it's being changed from the text input type to a non-text input type
+        if ($('#question-options').length < 1 && $type.val() !== 'text-input') {
+            addOptionsCreator($('#input-container')) // brings up the multiple choice options controls
+
+        } else if ($type.val() == 'text-input') { // checking that it's being changed to the text input type
+            $('#question-options').remove() //removes the multiple choice options controls
+        }
+        validateNewQuestion($('#question').val(), addQuestionButton, $type.val())
```
Comment (ArcticFoxJ) [inline public/js/surveyBuilderForm.js]: relying on lexical scope for addQuestionButton?

File: public/js/surveyBuilderForm.js [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+
+    var $type = $('#input-selector')
+
+    $type.change(function () { // attaching a listener for the question input type being changed
+
+        //checking that it's being changed from the text input type to a non-text input type
+        if ($('#question-options').length < 1 && $type.val() !== 'text-input') {
+            addOptionsCreator($('#input-container')) // brings up the multiple choice options controls
+
+        } else if ($type.val() == 'text-input') { // checking that it's being changed to the text input type
```
Comment (ArcticFoxJ) [inline public/js/surveyBuilderForm.js]: $type - relying on scope? possible to use $(this)?

File: public/js/surveyBuilderForm.js [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+
+    addQuestionButton.prop("disabled", true)
+
+    var $type = $('#input-selector')
+
+    $type.change(function () { // attaching a listener for the question input type being changed
+
+        //checking that it's being changed from the text input type to a non-text input type
+        if ($('#question-options').length < 1 && $type.val() !== 'text-input') {
+            addOptionsCreator($('#input-container')) // brings up the multiple choice options controls
```
Comment (ArcticFoxJ) [inline public/js/surveyBuilderForm.js]: should you comment a function call? surely the comment on the function should be enough?

File: public/js/surveyBuilderForm.js [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+        $(currentOption).parent('div').remove()
+        validateNewQuestion($('#question').val(), $('#add-question'), $('#input-selector').val())
+    }
+
+    /**
+     * Validates input of new question form
+     * Enables button if all inputs valid
+     *TODO put in the names of the parameters
+     * TODO change the data type of the 2nd parameter as it's a jQuery object, which is itself a JS object
+     * @param STRING question text in question's text box
```
Comment (JessI23) [inline public/js/surveyBuilderForm.js]: parameter text does not match parameter names in docblock

File: public/js/surveyBuilderForm.js (hunk 12 above)
Comment (ralph-rutter) [reply public/js/surveyBuilderForm.js]: Oh right, lol. I took a look and double space appears to be gone already.

File: public/js/surveyBuilderForm.js (hunk 16 above)
Comment (ralph-rutter) [reply public/js/surveyBuilderForm.js]: Cheers, I've removed this comment.

File: public/js/surveyBuilderForm.js [hunk 18]
```diff
@@ -190,4 +154,72 @@ $(function() {
...
+}
+
+/**
+ * creates question preview div using passed information and appends to container
+ *
+ * @param $container JQUERYSELECTOR container to append new question to
+ * @param question STRING question text
+ * @param response STRING html input(s) for question response
+ * @param $required JQUERYSELECTOR required checkbox
+ */
```
Comment (ralph-rutter) [inline public/js/surveyBuilderForm.js:208]: change the data type from JQUERYSELECTOR to OBJECT, as it is the selector's return value that we are assigning to the varibable, not the selector itself. jQuery selectors return jQuery objects, which are themselves JS objects.

**Pull request closed without merging**